annlite/core/codec/projector.py

```diff
@@ -0,0 +1,146 @@
+            data, then the more efficient ‘randomized’ method is enabled. Otherwise
+            the exact full SVD is computed and optionally truncated afterwards.
+
+            If full: run exact full SVD calling the standard LAPACK solver via scipy.
+            linalg.svd and select the components by postprocessing.
+
+            If arpack: run SVD truncated to n_components calling ARPACK solver via
+            scipy.sparse.linalg.svds. It requires strictly 0 < n_components < min(X.shape).
+    :param is_incremental: whether to use incremental projector, especially useful when
+            data size is too large to be loaded into memory at once.
+    :param batch_size: batch size used for incremental projector.
+    """
+
+    def __init__(
+        self,
+        n_components: int = 128,
+        whiten: Optional[bool] = False,
+        svd_solver: Optional[str] = 'auto',
+        is_incremental: Optional[bool] = False,
+        batch_size: Optional[int] = 512,
+    ):
+        super(ProjectorCodec, self).__init__(require_train=True)
+        self.n_components = n_components
+        self.whiten = whiten
+        self.svd_solver = svd_solver
+
+        self._is_trained = False
+
+        self.is_incremental = is_incremental
+        if not self.is_incremental:
+            from sklearn.decomposition import PCA
+
+            # fix the random seed to make sure that we can get the same result in each
+            # function call
+            self.pca = PCA(n_components=self.n_components, random_state=1234)
+        else:
+            from sklearn.decomposition import IncrementalPCA
+
+            self.batch_size = batch_size
+            self.pca = IncrementalPCA(
+                n_components=self.n_components, batch_size=self.batch_size
+            )
+
+    def fit(self, x: 'np.ndarray'):
+        """Train projector model
+
+        :param x: Training vectors with shape=(N, D)
+        """
+        assert x.dtype == np.float32
+        assert x.ndim == 2
+        assert (
+            x.shape[0] > self.n_components
+        ), 'number of input data must be larger than or equal to n_components'
+
+        self.pca.fit(x)
+        self._is_trained = True
+
+    def partial_fit(self, x: 'np.ndarray'):
+        """Given a batch of training vectors, update the internal projector.
+        This method is specially designed to be used when data does not fit in memory.
+
+        :param x: Training vectors with shape=(N, D)
+        """
+        assert x.dtype == np.float32
+        assert x.ndim == 2
+        assert (
+            x.shape[0] > self.n_components
+        ), 'number of input data must be larger than or equal to n_components'
+        assert (
+            len(x) % self.batch_size == 0
+        ), 'number of input data must be divided by batch size'
+
+        for i in range(0, len(x), self.batch_size):
+            self.pca.partial_fit(x[i : i + self.batch_size])
+        self._is_trained = True
```

Comment:
The input `x` of `partial_fit(x)` should be the partial training data, hence the proper usage of partial fit would look like:
```
def read_data(start, size):
    # read data from disk ...
    return x

for x in read_data(i, k):
    self.partial_fit(x)
self._is_train = True
```